Added a new constructor to create FF4J for a feature map

filippo22000 opened the pull request:
and Fixed missing spring boot tomcat version

clun commented:
Hi,

Thank you for your Pull request. 

Don't take it as personally but there are some element i would like to avoid and will make some adjustments.

- I worked hard to keep compatible with JDK1.6 as much as possible and upgrade only modules to jdk 1.7,8 when required. As a consequence the pom of ff4j-core should not be raised to 1.7 JDK or we should have to go 2.0 for next version.

- A constructor with Map would only work with InMemoryFeatureStore where there are dozens of stores implementation (in the constructor the store is not yet initialized and most of the time it's injected through Spring.). I propose you to have a method importFeatures(Map < String, Feature> ) on FF4J returning this or even better a Set of Feature to be able to write : 
```java
new FF4j().importFeatures(map);
new FF4j().importFeatures(set);
```

What do you think ?

filippo22000 commented:
Hi,
Thanks for your feedback.
I created that constructor because I have created several spring boot annotations with autoconfiguration that are using FF4J as feature toggling.

```
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.TYPE})
@Documented
@Import({FF4JRemoteAutoConfiguration.class})
public @interface EnableFF4JWithRemoteStore {
}
```

Using spring boot with these annotations we can create a FF4J consumers that create an inmemory instance from a map of features retrieved from a remote producer.

I used the constructor because in that scenario the store only exists if the map exists.

If you want to go with the _imporFeatures_ I suggest I create a builder like 
`FF4J.builder().features(map).build()`

What do you think?

For the Java version it s fine to me. Just consider that Java7 won't be supported anymore by Oracle soon.

Cheers,

Filippo

clun commented:
I like the builder idea, i will work on it.

"_we can create a FF4J consumers that create an inmemory instance from a map of features retrieved from a remote producer._"

I agree, this is the role of the [FF4jCacheProxy](https://github.com/clun/ff4j/blob/master/ff4j-core/src/main/java/org/ff4j/cache/FF4jCacheProxy.java) : keep a cache of feature on client side when store are remoted and handling TTL and different caching provider (in memory, redis, terracotta, JSR107).
[Sample use](https://github.com/clun/ff4j/blob/master/ff4j-core/src/test/java/org/ff4j/test/cache/CacheProxyTest.java)

Well,  I know, documentation is really **urgent** now....^_^

filippo22000 commented:
I just did it if you want  I push.
I introduced lombok and cleaned up a little bit the FF4J class
https://projectlombok.org/

Here the example

```
`package org.ff4j;

import static org.ff4j.audit.EventConstants.ACTION_CHECK_OFF;
import static org.ff4j.audit.EventConstants.ACTION_CHECK_OK;
import static org.ff4j.audit.EventConstants.SOURCE_JAVA;

/*
 * #%L
 * ff4j-core
 * %%
 * Copyright (C) 2013 Ff4J
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import java.io.IOException;
import java.io.InputStream;
import java.lang.reflect.InvocationTargetException;
import java.util.Map;
import java.util.Set;

import lombok.*;
import lombok.extern.slf4j.Slf4j;
import org.ff4j.audit.EventBuilder;
import org.ff4j.audit.EventPublisher;
import org.ff4j.audit.proxy.FeatureStoreAuditProxy;
import org.ff4j.audit.proxy.PropertyStoreAuditProxy;
import org.ff4j.audit.repository.EventRepository;
import org.ff4j.audit.repository.InMemoryEventRepository;
import org.ff4j.conf.XmlParser;
import org.ff4j.core.Feature;
import org.ff4j.core.FeatureStore;
import org.ff4j.core.FlippingExecutionContext;
import org.ff4j.core.FlippingStrategy;
import org.ff4j.exception.FeatureNotFoundException;
import org.ff4j.property.Property;
import org.ff4j.property.store.InMemoryPropertyStore;
import org.ff4j.property.store.PropertyStore;
import org.ff4j.security.AuthorizationsManager;
import org.ff4j.store.InMemoryFeatureStore;

/**
 * Principal class stands as public api to work with FF4J.
 *
 * <ul>
 * <p>
 * <li>
 * It embeddes a {@link FeatureStore} to record features statused. By default, features are stored into memory but you would like
 * to persist them in an external storage (as database) and choose among implementations available in different modules (jdbc,
 * mongo, http...).
 * </p>
 * 
 * <p>
 * <li>It embeddes a {@link AuthorizationsManager} to add permissions and limit usage of features to granted people. FF4J does not
 * created roles, it's rely on external security provider as SpringSecurity Apache Chiro.
 * </p>
 * 
 * <p>
 * <li>It embeddes a {@link EventRepository} to monitoring actions performed on features.
 * </p>
 * 
 * </ul>
 * 
 * @author Cedrick Lunven (@clunven)
 */
@Slf4j
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class FF4j {

    /** Do not through {@link FeatureNotFoundException} exception and but feature is required. */
    @Setter
    @Getter
    private boolean autocreate = false;
    
    /** Capture informations relative to audit. */
    @Getter
    private boolean enableAudit = false;

    /** Intialisation. */
    @Getter
    private final long startTime = System.currentTimeMillis();

    /** Version of ff4j. */
    @Getter
    private final String version = getClass().getPackage().getImplementationVersion();
    
    /** Source of initialization (JAVA_API, WEBAPI, SSH, CONSOLE...). */
    @Getter
    private final String source =  SOURCE_JAVA;
    
    /** Storage to persist feature within {@link FeatureStore}. */
    @Setter
    private FeatureStore featureStore = new InMemoryFeatureStore();
    
    /** Storage to persist properties within {@link PropertyStore}. */
    @Setter
    private PropertyStore propertiesStore = new InMemoryPropertyStore();

    /** Security policy to limit access through ACL with {@link AuthorizationsManager}. */
    @Getter
    @Setter
    private AuthorizationsManager authorizationsManager = null;

    /** Repository for audit event. */
    @Getter
    @Setter
    private EventRepository eventRepository = new InMemoryEventRepository();

    /** Event Publisher (threadpool, executor) to send data into {@link EventRepository} */
    @Setter
    private EventPublisher eventPublisher = null;
    
    private volatile boolean shutdownEventPublisher;
    
    /** Post Processing like audit enable. */
    private boolean initialized = false;

    /** Hold flipping execution context as Thread-safe data. */
    private ThreadLocal<FlippingExecutionContext> currentExecutionContext = new ThreadLocal<FlippingExecutionContext>();

    /**
     * This attribute indicates when call the alter bean throw de {@link InvocationTargetException}
     * or the wraps exception thrown by an invoked method or constructor
     */
    @Getter
    private boolean alterBeanThrowInvocationTargetException = true;
    


    /**
     * Constructor initializing ff4j with an InMemoryStore
     */
    public FF4j(String xmlFile) {
        this.featureStore = new InMemoryFeatureStore(xmlFile);
        this.propertiesStore = new InMemoryPropertyStore(xmlFile);
    }

    /**
     * Constructor initializing ff4j with an InMemoryStore using an InputStream. Simplify integration with Android through
     * <code>Asset</code>
     */
    public FF4j(InputStream xmlFileResourceAsStream) {
        this.featureStore = new InMemoryFeatureStore(xmlFileResourceAsStream);
    }



    /**
     * Ask if flipped.
     * 
     * @param featureID
     *            feature unique identifier.
     * @param executionContext
     *            current execution context
     * @return current feature status
     */
    public boolean check(String featureID) {
        return check(featureID, null);
    }

    /**
     * Elegant way to ask for flipping.
     * 
     * @param featureID
     *            feature unique identifier.
     * @param executionContext
     *            current execution context
     * @return current feature status
     */
    public boolean check(String featureID, FlippingExecutionContext executionContext) {
        Feature fp = getFeature(featureID);
        boolean flipped = fp.isEnable();

        // If authorization manager provided, apply security filter
        if (flipped && getAuthorizationsManager() != null) {
            flipped = flipped && isAllowed(fp);
        }

        // If custom strategy has been defined, delegate flipping to
        if (flipped && fp.getFlippingStrategy() != null) {
            flipped = flipped && fp.getFlippingStrategy().evaluate(featureID, getFeatureStore(), executionContext);
        }
        // Update current context
        currentExecutionContext.set(executionContext);
        
        // Any access is logged into audit system
        publishCheck(featureID, flipped);

        return flipped;
    }
    
    /**
     * Send target event to audit if expected.
     *
     * @param uid
     *      feature unique identifier
     * @param checked
     *      if the feature is checked or not
     */
    private void publishCheck(String uid, boolean checked) {
        if (isEnableAudit()) {
            getEventPublisher().publish(new EventBuilder(this)
                        .feature(uid)
                        .action(checked ? ACTION_CHECK_OK : ACTION_CHECK_OFF)
                        .build());
        }
    }

    /**
     * Overriding strategy on feature.
     * 
     * @param featureID
     *            feature unique identifier.
     * @param executionContext
     *            current execution context
     * @return
     */
    public boolean checkOveridingStrategy(String featureID, FlippingStrategy strats) {
        return checkOveridingStrategy(featureID, strats, currentExecutionContext.get());
    }

    /**
     * Overriding strategy on feature.
     * 
     * @param featureID
     *            feature unique identifier.
     * @param executionContext
     *            current execution context
     * @return
     */
    public boolean checkOveridingStrategy(String featureID, FlippingStrategy strats, FlippingExecutionContext executionContext) {
        Feature fp = getFeature(featureID);
        boolean flipped = fp.isEnable() && isAllowed(fp);
        if (strats != null) {
            flipped = flipped && strats.evaluate(featureID, getFeatureStore(), executionContext);
        }
        publishCheck(featureID, flipped);
        return flipped;
    }

    /**
     * Load SecurityProvider roles (e.g : SpringSecurity GrantedAuthorities)
     * 
     * @param featureName
     *            target name of the feature
     * @return if the feature is allowed
     */
    public boolean isAllowed(Feature featureName) {
        // No authorization manager, returning always true
        if (getAuthorizationsManager() == null) {
            return true;
        }
        // if no permissions, the feature is public
        if (featureName.getPermissions().isEmpty()) {
            return true;
        }
        Set<String> userRoles = getAuthorizationsManager().getCurrentUserPermissions();
        for (String expectedRole : featureName.getPermissions()) {
            if (userRoles.contains(expectedRole)) {
                return true;
            }
        }
        return false;
    }

    /**
     * Read Features from store.
     * 
     * @return get store features
     */
    public Map<String, Feature> getFeatures() {
        return getFeatureStore().readAll();
    }
    
    /**
     * Return all properties from store.
     *
     * @return
     * 		target property store.
     */
    public Map < String, Property<?>> getProperties() {
    	return getPropertiesStore().readAllProperties();
    }

    /**
     * Enable Feature.
     * 
     * @param featureID
     *            unique feature identifier.
     */
    public FF4j enable(String featureID) {
        try {
            getFeatureStore().enable(featureID);
        } catch (FeatureNotFoundException fnfe) {
            if (this.autocreate) {
                getFeatureStore().create(new Feature(featureID, true));
            } else {
            	throw fnfe;
            }
        }
        return this;
    }

    /**
     * Enable group.
     * 
     * @param groupName
     *            target groupeName
     * @return current instance
     */
    public FF4j enableGroup(String groupName) {
        getFeatureStore().enableGroup(groupName);
        return this;
    }

    /**
     * Disable group.
     * 
     * @param groupName
     *            target groupeName
     * @return current instance
     */
    public FF4j disableGroup(String groupName) {
        getFeatureStore().disableGroup(groupName);
        return this;
    }
    
    /**
     * Create new Feature.
     * 
     * @param featureID
     *            unique feature identifier.
     */
    public FF4j createFeature(Feature fp) {
        getFeatureStore().create(fp);
        return this;
    }
    
    
    /**
     * Create new Property.
     * 
     * @param featureID
     *            unique feature identifier.
     */
    public FF4j createProperty(Property<?> prop) {
        getPropertiesStore().createProperty(prop);
        return this;
    }
    
    /**
     * Create new Feature.
     * 
     * @param featureID
     *            unique feature identifier.
     */
    public FF4j createFeature(String featureName, boolean enable, String description) {
        return createFeature(new Feature(featureName, enable, description));
    }
    
    /**
     * Create new Feature.
     * 
     * @param featureID
     *            unique feature identifier.
     */
    public FF4j createFeature(String featureName, boolean enable) {
        return createFeature(featureName, enable, "");
    }
    
    /**
     * Create new Feature.
     * 
     * @param featureID
     *            unique feature identifier.
     */
    public FF4j createFeature(String featureName) {
        return createFeature(featureName, false, "");
    }

    /**
     * Disable Feature.
     * 
     * @param featureID
     *            unique feature identifier.
     */
    public FF4j disable(String featureID) {
        try {
            getFeatureStore().disable(featureID);
        } catch (FeatureNotFoundException fnfe) {
        	 if (this.autocreate) {
                 getFeatureStore().create(new Feature(featureID, false));
             } else {
             	throw fnfe;
             }
        }
        return this;
    }

    /**
     * Check if target feature exist.
     * 
     * @param featureId
     *            unique feature identifier.
     * @return flag to check existence of
     */
    public boolean exist(String featureId) {
        return getFeatureStore().exist(featureId);
    }

    /**
     * The feature will be create automatically if the boolea, autocreate is enabled.
     * 
     * @param featureID
     *            target feature ID
     * @return target feature.
     */
    public Feature getFeature(String featureID) {
        Feature fp = null;
        try {
            fp = getFeatureStore().read(featureID);
        } catch (FeatureNotFoundException fnfe) {
            if (this.autocreate) {
                fp = new Feature(featureID, false);
                getFeatureStore().create(fp);
            } else {
                throw fnfe;
            }
        }
        return fp;
    }

    /**
     * Export Feature through FF4J.
     * 
     * @return
     * @throws IOException
     */
    public InputStream exportFeatures() throws IOException {
        return new XmlParser().exportFeatures(getFeatureStore().readAll());
    }

    /**
     * Enable autocreation of features when not found.
     * 
     * @param flag
     *            target value for autocreate flag
     * @return current instance
     */
    public FF4j autoCreate(boolean flag) {
        setAutocreate(flag);
        return this;
    }
    
    /**
     * Enable autocreation of features when not found.
     * 
     * @param flag
     *            target value for autocreate flag
     * @return current instance
     */
    public FF4j autoCreate() {
        return autoCreate(true);
    }
    
    /**
     * Enable auditing of features when not found.
     * 
     * @param flag
     *            target value for autocreate flag
     * @return current instance
     */
    public FF4j audit() {
        return audit(true);
    }
            
    /**
     * Enable auditing of features when not found.
     * 
     * @param flag
     *            target value for autocreate flag
     * @return current instance
     */
    public FF4j audit(boolean val) {
         setEnableAudit(val);
         return this;
    }

    /**
     * Delete feature name.
     * 
     * @param fpId
     *            target feature
     */
    public FF4j delete(String fpId) {
        getFeatureStore().delete(fpId);
        return this;
    }
    
    /**
     * Delete new Property.
     * 
     * @param featureID
     *            unique feature identifier.
     */
    public FF4j deleteProperty(String propertyName) {
        getPropertiesStore().deleteProperty(propertyName);
        return this;
    }

    /** {@inheritDoc} */
    public String toString() {
        StringBuilder sb = new StringBuilder("{");
        long uptime = System.currentTimeMillis() - startTime;
        long daynumber = uptime / (1000 * 3600 * 24L);
        uptime = uptime - daynumber * 1000 * 3600 * 24L;
        long hourNumber = uptime / (1000 * 3600L);
        uptime = uptime - hourNumber * 1000 * 3600L;
        long minutenumber = uptime / (1000 * 60L);
        uptime = uptime - minutenumber * 1000 * 60L;
        long secondnumber = uptime / 1000L;
        sb.append("\"uptime\":\"");
        sb.append(daynumber + " day(s) ");
        sb.append(hourNumber + " hours(s) ");
        sb.append(minutenumber + " minute(s) ");
        sb.append(secondnumber + " seconds\"");
        sb.append(", \"autocreate\":" + isAutocreate());
        sb.append(", \"version\": \"" + version + "\"");
        
        // Display only if not null
        if (getFeatureStore() != null) {
            sb.append(", \"featuresStore\":");
            sb.append(getFeatureStore().toString());
        }
        if (getPropertiesStore() != null) {
            sb.append(", \"propertiesStore\":");
            sb.append(getPropertiesStore().toString());
        }
        if (getEventRepository() != null) {
            sb.append(", \"eventRepository\":");
            sb.append(getEventRepository().toString());
        }
        if (getAuthorizationsManager() != null) {
            sb.append(", \"authorizationsManager\":");
            sb.append(getAuthorizationsManager().toString());
        }
        sb.append("}");
        return sb.toString();
    }


    /**
     * Initialization of background components.
     */
    private synchronized void init() {
        
        // Execution Context
        FlippingExecutionContext context = new FlippingExecutionContext();
        this.currentExecutionContext.set(context);
        
        // Event Publisher
        eventPublisher = new EventPublisher(eventRepository);
        this.shutdownEventPublisher = true;
        
        // Audit is enabled, proxified for auditing
        if (isEnableAudit()) {
        	
        	if (featureStore != null && !(featureStore instanceof FeatureStoreAuditProxy)) {
                this.featureStore = new FeatureStoreAuditProxy(this,
                                                               featureStore);
            }
            if (propertiesStore != null && !(propertiesStore instanceof PropertyStoreAuditProxy)) {
                this.propertiesStore = new PropertyStoreAuditProxy(this,
                                                                   propertiesStore);
            }
        } else {
        	
        	 // Audit is disabled but could have been enabled before... removing PROXY if relevant
        	 if (featureStore != null && featureStore instanceof FeatureStoreAuditProxy) {
        		 this.featureStore = ((FeatureStoreAuditProxy) featureStore).getTarget();
        	 }
        	 if (propertiesStore != null && propertiesStore instanceof PropertyStoreAuditProxy) {
        		 this.propertiesStore = ((PropertyStoreAuditProxy) propertiesStore).getTarget();
             }
        }
        
        // Flag as OK
        this.initialized = true;
    }
    
    /**
     * Access store as static way (single store).
     * 
     * @return current store
     */
    public FeatureStore getFeatureStore() {
        if (!initialized) init();
        return featureStore;
    }
    
    
    /**
     * Getter accessor for attribute 'eventPublisher'.
     * 
     * @return current value of 'eventPublisher'
     */
    public synchronized EventPublisher getEventPublisher() {
        if (!initialized) init();
        return eventPublisher;
    }
    
    /**
     * Getter accessor for attribute 'propertiesStore'.
     *
     * @return
     *       current value of 'propertiesStore'
     */
    public PropertyStore getPropertiesStore() {
        if (!initialized) init();
        return propertiesStore;
    }
    
    /**
     * Initialize flipping execution context.
     *
     * @return
     *      get current context
     */
    public FlippingExecutionContext getCurrentContext() {
        if (!initialized) init();
        
        if (null == this.currentExecutionContext.get()) {
            this.currentExecutionContext.set(new FlippingExecutionContext());
        }
        return this.currentExecutionContext.get();
    }






    /**
     * Clear context.
     */
    public void removeCurrentContext() {
        this.currentExecutionContext.remove();
    }


    /**
     * Setter accessor for attribute 'enableAudit'.
     *
     * @param enableAudit
     * 		new value for 'enableAudit '
     */
    public void setEnableAudit(boolean enableAudit) {
    	this.enableAudit = enableAudit;
    	
    	// if you disable the audit you should remove the auditProxy
    	initialized = false;
    }
    
    /**
     * Required for spring namespace and 'fileName' attribut on ff4j tag.
     *
     * @param fname
     *      target name
     */
    public void setFileName(String fname) { }
    public void setAuthManager(String mnger) { }

    /**
     * Shuts down the event publisher if we actually started it (As opposed to
     * having it dependency-injected).
     */
    public void stop() {
        if (this.eventPublisher != null && this.shutdownEventPublisher) {
            this.eventPublisher.stop();
        }
    }


        /**
     * Enable Alter bean Throw InvocationTargetException, when enabled
     * the alter bean method always throw {@link InvocationTargetException}
     */
    public void enableAlterBeanThrowInvocationTargetException() {
        this.alterBeanThrowInvocationTargetException = true;
    }

    /**
     * Disable Alter bean Throw InvocationTargetException, when disabled
     * the alter bean method always throw the exception cause.
     */
    public void disableAlterBeanThrowInvocationTargetException() {
        this.alterBeanThrowInvocationTargetException = false;
    }


}
`
```

And here how to use it

                 
                       ```
                          FF4j ff4j = new FF4j().builder()
							  .featureStore(new InMemoryFeatureStore(featureMap))
							  .build();
```


As you can see with lombok you reduce drastically the code making it more readable.

Just let me know if you want me to push.

Cheers,

Filippo

clun commented:
Hi, i will push tonight

`Lombok` is pretty powerful framework I agree. But look at the dependencies of `ff4j-core`, even if hard to believe : **there are none** *(except for tests)*. In the first version i relied on `slf4j` but even this one is not really compliant with Android :) so I removed.

filippo22000 commented:
Ok I understand.
If you install the intellij/exclipse plugin you can de-lombok and you will
get all your builders, getters and setters done in one sec
Cheers,
Filippo

On 11 May 2016 at 09:53, Cédrick Lunven <notifications@github.com> wrote:

> Hi, i will push tonight
>
> Lombok is pretty powerful framework I agree. But look at the dependencies
> of ff4j-core, even if hard to believe : *there are none* *(except for
> tests)*. In the first version i relied on slf4j but even this one is not
> really compliant with Android :) so I removed.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/clun/ff4j/pull/123#issuecomment-218400457>
>

clun commented:
Added the methods to FF4J and each store in a builder pattern (return this);

```java
public FF4j importFeatures(Collection < Feature> features);
public FF4j importProperties(Collection < Property<?> props);
```

Sample use:
```java
new FF4j().importFeatures(null).importProperties(null);
```
